### PR TITLE
Revert breaking type changes.

### PR DESCRIPTION
**Pull request by markstory**
As per #12335 this reverts breaking changes made in #12286

### CODE REVIEW

**Review by lorenzo — APPROVED** [review]

File: src/Filesystem/File.php [hunk 1]
```diff
@@ -450,29 +450,29 @@ public function exists(): bool
     /**
      * Returns the "chmod" (permissions) of the file.
      *
-     * @return string|null Permissions for the file, or null in case of an error
+     * @return string|bool Permissions for the file, or false in case of an error
```
Comment (dereuromark) [inline src/Filesystem/File.php]: bool must be false here
Same for the other cases, as true can never be returned (we only abuse false here for null).

**Review by dereuromark — CHANGES_REQUESTED** [review]

File: src/Filesystem/File.php (hunk 1 above)
Comment (markstory) [reply src/Filesystem/File.php]: PHPStan doesn't seem to have blown up so far. Why would `bool` cause issues?

**Comment by markstory** [comment]
@dereuromark Yes I think we can schedule replacing the bool|other changes for 5.0. Another option would be to replace the 'Filesystem' package with an improved implementation avoiding the need to break compatibility in a subtle way.

File: src/Filesystem/File.php (hunk 1 above)
Comment (markstory) [reply src/Filesystem/File.php]: Ok. We use `bool` quite extensively as it is a type while `false` is not. If tooling works better with `string|false` that sounds like a separate change.

**Review by dereuromark — APPROVED** [review]

**Pull request merged**